diff --git a/mos6502.cpp b/mos6502.cpp
index e1e93e5..fcbc574 100644
--- a/mos6502.cpp
+++ b/mos6502.cpp
@@ -29,15 +29,14 @@
 
 mos6502::Instr mos6502::InstrTable[256];
 
-mos6502::mos6502(BusRead r, BusWrite w, ClockCycle c)
+mos6502::mos6502(i_memory_access *memory_access, ClockCycle c)
 	: reset_A(0x00)
     , reset_X(0x00)
     , reset_Y(0x00)
     , reset_sp(0xFD)
     , reset_status(CONSTANT)
 {
-	Write = (BusWrite)w;
-	Read = (BusRead)r;
+	this->memory_access = memory_access;
 	Cycle = (ClockCycle)c;
 
 	static bool initialized = false;
@@ -734,8 +733,8 @@ uint16_t mos6502::Addr_ABS()
 	uint16_t addrH;
 	uint16_t addr;
 
-	addrL = Read(pc++);
-	addrH = Read(pc++);
+	addrL = memory_access->read(pc++);
+	addrH = memory_access->read(pc++);
 
 	addr = addrL + (addrH << 8);
 
@@ -744,7 +743,7 @@ uint16_t mos6502::Addr_ABS()
 
 uint16_t mos6502::Addr_ZER()
 {
-	return Read(pc++);
+	return memory_access->read(pc++);
 }
 
 uint16_t mos6502::Addr_IMP()
@@ -757,7 +756,7 @@ uint16_t mos6502::Addr_REL()
 	uint16_t offset;
 	uint16_t addr;
 
-	offset = (uint16_t)Read(pc++);
+	offset = (uint16_t)memory_access->read(pc++);
 	if (offset & 0x80) offset |= 0xFF00;
 	addr = pc + (int16_t)offset;
 	return addr;
@@ -772,17 +771,17 @@ uint16_t mos6502::Addr_ABI()
 	uint16_t abs;
 	uint16_t addr;
 
-	addrL = Read(pc++);
-	addrH = Read(pc++);
+	addrL = memory_access->read(pc++);
+	addrH = memory_access->read(pc++);
 
 	abs = (addrH << 8) | addrL;
 
-	effL = Read(abs);
+	effL = memory_access->read(abs);
 
 #ifndef CMOS_INDIRECT_JMP_FIX
-	effH = Read((abs & 0xFF00) + ((abs + 1) & 0x00FF) );
+	effH = memory_access->read((abs & 0xFF00) + ((abs + 1) & 0x00FF) );
 #else
-	effH = Read(abs + 1);
+	effH = memory_access->read(abs + 1);
 #endif
 
 	addr = effL + 0x100 * effH;
@@ -792,13 +791,13 @@ uint16_t mos6502::Addr_ABI()
 
 uint16_t mos6502::Addr_ZEX()
 {
-	uint16_t addr = (Read(pc++) + X) & 0xFF;
+	uint16_t addr = (memory_access->read(pc++) + X) & 0xFF;
 	return addr;
 }
 
 uint16_t mos6502::Addr_ZEY()
 {
-	uint16_t addr = (Read(pc++) + Y) & 0xFF;
+	uint16_t addr = (memory_access->read(pc++) + Y) & 0xFF;
 	return addr;
 }
 
@@ -808,8 +807,8 @@ uint16_t mos6502::Addr_ABX()
 	uint16_t addrL;
 	uint16_t addrH;
 
-	addrL = Read(pc++);
-	addrH = Read(pc++);
+	addrL = memory_access->read(pc++);
+	addrH = memory_access->read(pc++);
 
 	addr = addrL + (addrH << 8) + X;
 	return addr;
@@ -821,8 +820,8 @@ uint16_t mos6502::Addr_ABY()
 	uint16_t addrL;
 	uint16_t addrH;
 
-	addrL = Read(pc++);
-	addrH = Read(pc++);
+	addrL = memory_access->read(pc++);
+	addrH = memory_access->read(pc++);
 
 	addr = addrL + (addrH << 8) + Y;
 	return addr;
@@ -835,9 +834,9 @@ uint16_t mos6502::Addr_INX()
 	uint16_t zeroH;
 	uint16_t addr;
 
-	zeroL = (Read(pc++) + X) & 0xFF;
+	zeroL = (memory_access->read(pc++) + X) & 0xFF;
 	zeroH = (zeroL + 1) & 0xFF;
-	addr = Read(zeroL) + (Read(zeroH) << 8);
+	addr = memory_access->read(zeroL) + (memory_access->read(zeroH) << 8);
 
 	return addr;
 }
@@ -848,9 +847,9 @@ uint16_t mos6502::Addr_INY()
 	uint16_t zeroH;
 	uint16_t addr;
 
-	zeroL = Read(pc++);
+	zeroL = memory_access->read(pc++);
 	zeroH = (zeroL + 1) & 0xFF;
-	addr = Read(zeroL) + (Read(zeroH) << 8) + Y;
+	addr = memory_access->read(zeroL) + (memory_access->read(zeroH) << 8) + Y;
 
 	return addr;
 }
@@ -862,8 +861,8 @@ void mos6502::Reset()
 	X = reset_X;
 
 	// load PC from reset vector
-	uint8_t pcl = Read(rstVectorL);
-	uint8_t pch = Read(rstVectorH);
+	uint8_t pcl = memory_access->read_raw(rstVectorL);
+	uint8_t pch = memory_access->read_raw(rstVectorH);
 	pc = (pch << 8) + pcl;
 
 	sp = reset_sp;
@@ -877,7 +876,7 @@ void mos6502::Reset()
 
 void mos6502::StackPush(uint8_t byte)
 {
-	Write(0x0100 + sp, byte);
+	memory_access->write(0x0100 + sp, byte);
 	if(sp == 0x00) sp = 0xFF;
 	else sp--;
 }
@@ -886,7 +885,7 @@ uint8_t mos6502::StackPop()
 {
 	if(sp == 0xFF) sp = 0x00;
 	else sp++;
-	return Read(0x0100 + sp);
+	return memory_access->read(0x0100 + sp);
 }
 
 void mos6502::IRQ()
@@ -900,8 +899,8 @@ void mos6502::IRQ()
 		SET_INTERRUPT(1);
 
 		// load PC from interrupt request vector
-		uint8_t pcl = Read(irqVectorL);
-		uint8_t pch = Read(irqVectorH);
+		uint8_t pcl = memory_access->read(irqVectorL);
+		uint8_t pch = memory_access->read(irqVectorH);
 		pc = (pch << 8) + pcl;
 	}
 	return;
@@ -916,8 +915,8 @@ void mos6502::NMI()
 	SET_INTERRUPT(1);
 
 	// load PC from non-maskable interrupt vector
-	uint8_t pcl = Read(nmiVectorL);
-	uint8_t pch = Read(nmiVectorH);
+	uint8_t pcl = memory_access->read(nmiVectorL);
+	uint8_t pch = memory_access->read(nmiVectorH);
 	pc = (pch << 8) + pcl;
 	return;
 }
@@ -933,7 +932,7 @@ void mos6502::Run(
 	while(cyclesRemaining > 0 && !illegalOpcode)
 	{
 		// fetch
-		opcode = Read(pc++);
+		opcode = memory_access->read(pc++);
 
 		// decode
 		instr = InstrTable[opcode];
@@ -960,7 +959,7 @@ void mos6502::RunEternally()
 	while(!illegalOpcode)
 	{
 		// fetch
-		opcode = Read(pc++);
+		opcode = memory_access->read(pc++);
 
 		// decode
 		instr = InstrTable[opcode];
@@ -1069,7 +1068,7 @@ void mos6502::Op_ILLEGAL(uint16_t src)
 
 void mos6502::Op_ADC(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	unsigned int tmp = m + A + (IF_CARRY() ? 1 : 0);
 	SET_ZERO(!(tmp & 0xFF));
 	if (IF_DECIMAL())
@@ -1098,7 +1097,7 @@ void mos6502::Op_ADC(uint16_t src)
 
 void mos6502::Op_AND(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	uint8_t res = m & A;
 	SET_NEGATIVE(res & 0x80);
 	SET_ZERO(!res);
@@ -1109,13 +1108,13 @@ void mos6502::Op_AND(uint16_t src)
 
 void mos6502::Op_ASL(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	SET_CARRY(m & 0x80);
 	m <<= 1;
 	m &= 0xFF;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
-	Write(src, m);
+	memory_access->write_data(src, m);
 	return;
 }
 
@@ -1161,7 +1160,7 @@ void mos6502::Op_BEQ(uint16_t src)
 
 void mos6502::Op_BIT(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	uint8_t res = m & A;
 	SET_NEGATIVE(res & 0x80);
 	status = (status & 0x3F) | (uint8_t)(m & 0xC0) | CONSTANT | BREAK;
@@ -1203,7 +1202,7 @@ void mos6502::Op_BRK(uint16_t src)
 	StackPush(pc & 0xFF);
 	StackPush(status | CONSTANT | BREAK);
 	SET_INTERRUPT(1);
-	pc = (Read(irqVectorH) << 8) + Read(irqVectorL);
+	pc = (memory_access->read_data(irqVectorH) << 8) + memory_access->read_data(irqVectorL);
 	return;
 }
 
@@ -1251,7 +1250,7 @@ void mos6502::Op_CLV(uint16_t src)
 
 void mos6502::Op_CMP(uint16_t src)
 {
-	unsigned int tmp = A - Read(src);
+	unsigned int tmp = A - memory_access->read_data(src);
 	SET_CARRY(tmp < 0x100);
 	SET_NEGATIVE(tmp & 0x80);
 	SET_ZERO(!(tmp & 0xFF));
@@ -1260,7 +1259,7 @@ void mos6502::Op_CMP(uint16_t src)
 
 void mos6502::Op_CPX(uint16_t src)
 {
-	unsigned int tmp = X - Read(src);
+	unsigned int tmp = X - memory_access->read_data(src);
 	SET_CARRY(tmp < 0x100);
 	SET_NEGATIVE(tmp & 0x80);
 	SET_ZERO(!(tmp & 0xFF));
@@ -1269,7 +1268,7 @@ void mos6502::Op_CPX(uint16_t src)
 
 void mos6502::Op_CPY(uint16_t src)
 {
-	unsigned int tmp = Y - Read(src);
+	unsigned int tmp = Y - memory_access->read_data(src);
 	SET_CARRY(tmp < 0x100);
 	SET_NEGATIVE(tmp & 0x80);
 	SET_ZERO(!(tmp & 0xFF));
@@ -1278,11 +1277,11 @@ void mos6502::Op_CPY(uint16_t src)
 
 void mos6502::Op_DEC(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	m = (m - 1) & 0xFF;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
-	Write(src, m);
+	memory_access->write_data(src, m);
 	return;
 }
 
@@ -1308,7 +1307,7 @@ void mos6502::Op_DEY(uint16_t src)
 
 void mos6502::Op_EOR(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	m = A ^ m;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
@@ -1317,11 +1316,11 @@ void mos6502::Op_EOR(uint16_t src)
 
 void mos6502::Op_INC(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	m = (m + 1) & 0xFF;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
-	Write(src, m);
+	memory_access->write_data(src, m);
 }
 
 void mos6502::Op_INX(uint16_t src)
@@ -1357,7 +1356,7 @@ void mos6502::Op_JSR(uint16_t src)
 
 void mos6502::Op_LDA(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
 	A = m;
@@ -1365,7 +1364,7 @@ void mos6502::Op_LDA(uint16_t src)
 
 void mos6502::Op_LDX(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
 	X = m;
@@ -1373,7 +1372,7 @@ void mos6502::Op_LDX(uint16_t src)
 
 void mos6502::Op_LDY(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
 	Y = m;
@@ -1381,12 +1380,12 @@ void mos6502::Op_LDY(uint16_t src)
 
 void mos6502::Op_LSR(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	SET_CARRY(m & 0x01);
 	m >>= 1;
 	SET_NEGATIVE(0);
 	SET_ZERO(!m);
-	Write(src, m);
+	memory_access->write_data(src, m);
 }
 
 void mos6502::Op_LSR_ACC(uint16_t src)
@@ -1406,7 +1405,7 @@ void mos6502::Op_NOP(uint16_t src)
 
 void mos6502::Op_ORA(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	m = A | m;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
@@ -1442,14 +1441,14 @@ void mos6502::Op_PLP(uint16_t src)
 
 void mos6502::Op_ROL(uint16_t src)
 {
-	uint16_t m = Read(src);
+	uint16_t m = memory_access->read_data(src);
 	m <<= 1;
 	if (IF_CARRY()) m |= 0x01;
 	SET_CARRY(m > 0xFF);
 	m &= 0xFF;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
-	Write(src, m);
+	memory_access->write_data(src, m);
 	return;
 }
 
@@ -1468,14 +1467,14 @@ void mos6502::Op_ROL_ACC(uint16_t src)
 
 void mos6502::Op_ROR(uint16_t src)
 {
-	uint16_t m = Read(src);
+	uint16_t m = memory_access->read_data(src);
 	if (IF_CARRY()) m |= 0x100;
 	SET_CARRY(m & 0x01);
 	m >>= 1;
 	m &= 0xFF;
 	SET_NEGATIVE(m & 0x80);
 	SET_ZERO(!m);
-	Write(src, m);
+	memory_access->write_data(src, m);
 	return;
 }
 
@@ -1518,7 +1517,7 @@ void mos6502::Op_RTS(uint16_t src)
 
 void mos6502::Op_SBC(uint16_t src)
 {
-	uint8_t m = Read(src);
+	uint8_t m = memory_access->read_data(src);
 	unsigned int tmp = A - m - (IF_CARRY() ? 0 : 1);
 	SET_NEGATIVE(tmp & 0x80);
 	SET_ZERO(!(tmp & 0xFF));
@@ -1557,19 +1556,19 @@ void mos6502::Op_SEI(uint16_t src)
 
 void mos6502::Op_STA(uint16_t src)
 {
-	Write(src, A);
+	memory_access->write_data(src, A);
 	return;
 }
 
 void mos6502::Op_STX(uint16_t src)
 {
-	Write(src, X);
+	memory_access->write_data(src, X);
 	return;
 }
 
 void mos6502::Op_STY(uint16_t src)
 {
-	Write(src, Y);
+	memory_access->write_data(src, Y);
 	return;
 }
 
diff --git a/mos6502.h b/mos6502.h
index b976c22..39ec0a9 100644
--- a/mos6502.h
+++ b/mos6502.h
@@ -8,10 +8,11 @@
 
 #pragma once
 #include <stdint.h>
+#include "emulation/memory_access.hpp"
 
 class mos6502
 {
-private:
+protected:
     // register reset values
     uint8_t reset_A;
     uint8_t reset_X;
@@ -142,24 +143,22 @@ private:
 	static const uint16_t nmiVectorH = 0xFFFB;
 	static const uint16_t nmiVectorL = 0xFFFA;
 
-	// read/write/clock-cycle callbacks
-	typedef void (*BusWrite)(uint16_t, uint8_t);
-	typedef uint8_t (*BusRead)(uint16_t);
+	// clock-cycle callbacks
 	typedef void (*ClockCycle)(mos6502*);
-	BusRead Read;
-	BusWrite Write;
 	ClockCycle Cycle;
 
 	// stack operations
-	inline void StackPush(uint8_t byte);
-	inline uint8_t StackPop();
+	void StackPush(uint8_t byte);
+	uint8_t StackPop();
+
+	i_memory_access *memory_access;
 
 public:
 	enum CycleMethod {
 		INST_COUNT,
 		CYCLE_COUNT,
 	};
-	mos6502(BusRead r, BusWrite w, ClockCycle c = nullptr);
+	mos6502(i_memory_access *i_memory_access, ClockCycle c = nullptr);
 	void NMI();
 	void IRQ();
 	void Reset();
